src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -1002,6 +1027,30 @@ internal JsonPropertyDictionary<JsonPropertyInfo> CreatePropertyCache(int capaci
             return new JsonPropertyDictionary<JsonPropertyInfo>(Options.PropertyNameCaseInsensitive, capacity);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void InitializeRequiredProperties(ref ReadStack state)
+        {
+            if (RequiredProperties != null)
+            {
+                Debug.Assert(state.Current.RequiredPropertiesLeft == null);
+                state.Current.RequiredPropertiesLeft = new HashSet<JsonPropertyInfo>(RequiredProperties);
```

Comment:
What's are the equality semantics for `JsonPropertyInfo`? I'm guessing it's reference equality which happens to work, but I think that's predicated on us not implementing equality in the future and not making it possible to add the same `JsonPropertyInfo` instance twice in the properties list (because it would conflict on the requirement for `Name` to be unique). 

I would recommend using `HashSet<string>` keyed on the property names just to make things more explicit.

Reply:
I think having it keyed on the reference is faster than on reference and because they have to be unique (as you mentioned because of Name having to be unique). On the other hand string has to be compared everytime and in some cases we'd require to do UTF-8 transcoding just to be able to add it to HashSet

Reply:
Maybe, but I still think the fact that this works is accidental and depends on a couple of unrelated invariants. I think it's best to be deliberate here. If keying on the `Name` string proves to incur significant perf overhead, one alternative might be to key on the `int` index in the list of properties, since at this point it has become immutable.

Reply:
I like the idea of using index

Reply:
See also https://github.com/dotnet/runtime/pull/72937#discussion_r931928261. Might make sense to wrap inside a bespoke type (which could be a struct) whose surface APIs take `JsonPropertyInfo`.